File: android/src/main/java/com/apppresser/onesignal/CapOneSignalPlugin.java
package com.apppresser.onesignal;

import android.Manifest;
import com.getcapacitor.JSObject;
import com.getcapacitor.Logger;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;

import android.content.pm.PackageManager;
import android.os.Build;
import com.getcapacitor.PermissionState;
import com.getcapacitor.annotation.Permission;
import com.getcapacitor.annotation.PermissionCallback;
import com.onesignal.OneSignal;
import com.onesignal.debug.LogLevel;

@CapacitorPlugin(
    name = "CapOneSignal",
    permissions = { @Permission(alias = "notification", strings = { Manifest.permission.POST_NOTIFICATIONS }) }
)
public class CapOneSignalPlugin extends Plugin {

    private CapOneSignal implementation = new CapOneSignal();

    @PluginMethod
    public void initialize(PluginCall call) {
        String appId = call.getString("appID");
        if (appId == null || appId.isEmpty()) {
            call.reject("appID is required");
            return;
        }

        // Initialize OneSignal via implementation helper
        implementation.initialize(getContext(), appId);

        call.resolve();
    }

    @PluginMethod
    public void requestPermission(PluginCall call) {
        // For Android 13 (API 33) and newer, we always trigger the runtime permission prompt.
        // If the permission has already been granted, Capacitor will resolve successfully.
        // This ensures the dialog is shown for new installations, and for existing ones
        // where the user may not have been prompted before.
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
            requestPermissionForAlias("notification", call, "notificationPermissionCallback");
        } else {
            // For older Android versions, use the legacy implementation
            boolean fallbackToSettings = call.getBoolean("fallbackToSettings", false);
            implementation.requestPermission(getContext(), call, fallbackToSettings);

            // Re-check the status after the call, as the user might have been sent to settings
            boolean isEnabled = implementation.areNotificationsEnabled(getContext());
            JSObject result = new JSObject();
            result.put("accepted", isEnabled);
            call.resolve(result);
        }
    }

    @PermissionCallback
    private void notificationPermissionCallback(PluginCall call) {
        JSObject result = new JSObject();
        if (getPermissionState("notification") == PermissionState.GRANTED) {
            Logger.info(getLogTag(), "User granted notification permission.");
            result.put("accepted", true);
            call.resolve(result);
        } else {
            Logger.warn(getLogTag(), "User denied notification permission.");
            // You can either reject or resolve with accepted: false
            result.put("accepted", false);
            call.resolve(result);
        }
    }

    @PluginMethod
    public void setExternalUserId(PluginCall call) {
        String userId = call.getString("userID");
        if (userId == null || userId.isEmpty()) {
            call.reject("userID is required");
            return;
        }

        implementation.setExternalUserId(userId);

        call.resolve();
    }

    @PluginMethod
    public void setLogLevel(PluginCall call) {
        String level = call.getString("level");
        if (level == null || level.isEmpty()) {
            call.reject("level is required");
            return;
        }

        implementation.setLogLevel(level);

        call.resolve();
    }

    @PluginMethod
    public void clearExternalUserId(PluginCall call) {
        implementation.clearExternalUserId();

        call.resolve();
    }

    @PluginMethod
    public void addTags(PluginCall call) {
        JSObject tagsObject = call.getObject("tags");
        if (tagsObject == null) {
            call.reject("tags is required");
            return;
        }

        java.util.Map<String, String> tags = new java.util.HashMap<>();
        java.util.Iterator<String> keys = tagsObject.keys();
        while (keys.hasNext()) {
            String key = keys.next();
            String value = tagsObject.optString(key, null);
            if (value != null) {
                tags.put(key, value);
            }
        }

        if (tags.isEmpty()) {
            call.reject("tags must be an object with string values");
            return;
        }

        implementation.addTags(tags);
        call.resolve();
    }

    @PluginMethod
    public void removeTags(PluginCall call) {
        JSObject tagsObject = call.getObject("tags");
        if (tagsObject == null) {
            call.reject("tags is required");
            return;
        }

        java.util.List<String> tagKeys = new java.util.ArrayList<>();
        java.util.Iterator<String> keys = tagsObject.keys();
        while (keys.hasNext()) {
            tagKeys.add(keys.next());
        }

        if (tagKeys.isEmpty()) {
            call.reject("tags must be an object with keys to remove");
            return;
        }

        implementation.removeTags(tagKeys);
        call.resolve();
    }

    @PluginMethod
    public void addTag(PluginCall call) {
        String tagKey = call.getString("tagKey");
        if (tagKey == null || tagKey.isEmpty()) {
            call.reject("tagKey is required");
            return;
        }

        String tagValue = call.getString("tagValue");
        if (tagValue == null || tagValue.isEmpty()) {
            call.reject("tagValue is required");
            return;
        }

        implementation.addTag(tagKey, tagValue);
        call.resolve();
    }

    @PluginMethod
    public void removeTag(PluginCall call) {
        String tag = call.getString("tag");
        if (tag == null || tag.isEmpty()) {
            call.reject("tag is required");
            return;
        }

        implementation.removeTag(tag);
        call.resolve();
    }

}

File: ios/Sources/CapOneSignalPlugin/CapOneSignalPlugin.swift
import Capacitor
import Foundation

/// Please read the Capacitor iOS Plugin Development Guide
/// here: https://capacitorjs.com/docs/plugins/ios
@objc(CapOneSignalPlugin)
public class CapOneSignalPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "CapOneSignalPlugin"
    public let jsName = "CapOneSignal"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "initialize", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "requestPermission", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setLogLevel", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "setExternalUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "clearExternalUserId", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeTags", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "addTag", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "removeTag", returnType: CAPPluginReturnPromise),
    ]
    private let implementation = CapOneSignal()

    @objc func initialize(_ call: CAPPluginCall) {
        guard let appID = call.getString("appID") else {
            call.reject("appId is required")
            return
        }
        implementation.initialize(appID)
        call.resolve()
    }

    @objc func setLogLevel(_ call: CAPPluginCall) {
        guard let level = call.getString("level") else {
            call.reject("level is required")
            return
        }
        implementation.setLogLevel(level)
        call.resolve()
    }

    @objc func requestPermission(_ call: CAPPluginCall) {
        // iOS permission request logic - resolve with whether user accepted notifications
        implementation.requestPermission { accepted in
            call.resolve([
                "accepted": accepted
            ])
        }
    }

    @objc func setExternalUserId(_ call: CAPPluginCall) {
        guard let userID = call.getString("userID") else {
            call.reject("externalUserId is required")
            return
        }
        implementation.setExternalUserId(userID)
        call.resolve()
    }

    @objc func clearExternalUserId(_ call: CAPPluginCall) {
        implementation.clearExternalUserId()
        call.resolve()
    }

    @objc func addTags(_ call: CAPPluginCall) {
        guard let tagsAny = call.getObject("tags") as? [String: Any] else {
            call.reject("tags is required")
            return
        }
        let tags = tagsAny.compactMapValues { ($0 as? String) }
        if tags.isEmpty {
            call.reject("tags must be a [String: Any] dictionary")
            return
        }
        implementation.addTags(tags) { success in
            if success {
                call.resolve()
            } else {
                call.reject("Failed to add tags")
            }
        }
    }

    @objc func removeTags(_ call: CAPPluginCall) {
        guard let tagsAny = call.getObject("tags") as? [String: Any] else {
            call.reject("tags is required")
            return
        }
        let tagKeys = tagsAny.keys.map { String($0) }
        if tagKeys.isEmpty {
            call.reject("tags must be a [String: Any] dictionary with keys to remove")
            return
        }
        implementation.removeTags(tagKeys) { success in
            if success {
                call.resolve()
            } else {
                call.reject("Failed to remove tags")
            }
        }
    }

    @objc func addTag(_ call: CAPPluginCall) {
        guard let tagKey = call.getString("tagKey") else {
            call.reject("tagKey is required")
            return
        }
        guard let tagValue = call.getString("tagValue") else {
            call.reject("tagValue is required")
            return
        }
        implementation.sendTag(tagKey, tagValue) { success in
            if success {
                call.resolve()
            } else {
                call.reject("Failed to add tag")
            }
        }
    }

    @objc func removeTag(_ call: CAPPluginCall) {
        guard let tag = call.getString("tag") else {
            call.reject("tag is required")
            return
        }
        implementation.removeTag(tag) { success in
            if success {
                call.resolve()
            } else {
                call.reject("Failed to remove tag")
            }
        }
    }

}
